src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs

```diff
@@ -107,10 +162,165 @@ private string CreateVaryByCacheKey(CachedVaryBy varyBy)
+            var request = HttpContext.Request;
+            if (string.Equals("GET", request.Method, StringComparison.OrdinalIgnoreCase))
+            {
+                _responseType = ResponseType.FullReponse;
+            }
+            else if (string.Equals("HEAD", request.Method, StringComparison.OrdinalIgnoreCase))
+            {
+                _responseType = ResponseType.HeadersOnly;
+            }
+            else
+            {
+                return false;
+            }
+
+            // Verify existence of authorization headers
+            // TODO: The server may indicate that the response to these request are cacheable
+            if (!string.IsNullOrEmpty(request.Headers[HeaderNames.Authorization]))
+            {
+                return false;
+            }
+
+            // Verify request cache-control parameters
+            // TODO: no-cache requests can be retrieved upon validation with origin
+            if (!string.IsNullOrEmpty(request.Headers[HeaderNames.CacheControl]))
+            {
+                if (RequestCacheControl.NoCache || RequestCacheControl.NoStore)
+                {
+                    return false;
+                }
+            }
+            else
+            {
+                // Support for legacy HTTP 1.0 cache directive
+                var pragmaHeaderValues = request.Headers[HeaderNames.Pragma];
+                foreach (var directive in pragmaHeaderValues)
+                {
+                    if (string.Equals("no-cache", directive, StringComparison.OrdinalIgnoreCase))
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            // TODO: Verify global middleware settings? Explicit ignore list, range requests, etc.
+            return true;
+        }
+
+        internal bool ResponseIsCacheable()
+        {
+            // Only cache pages explicitly marked with public
+            // TODO: Consider caching responses that are not marked as public but otherwise cacheable?
+            if (!ResponseCacheControl.Public)
+            {
+                return false;
+            }
+
+            // Check no-store
+            if (ResponseCacheControl.NoStore)
+            {
+                return false;
+            }
+
+            // Check no-cache
+            // TODO: Handle no-cache with headers
+            if (ResponseCacheControl.NoCache)
+            {
+                return false;
+            }
+
+            var response = HttpContext.Response;
+
+            // Do not cache responses varying by *
+            if (string.Equals(response.Headers[HeaderNames.Vary], "*", StringComparison.OrdinalIgnoreCase))
+            {
+                return false;
+            }
+
+            // Do not cache responses that may bypass the response body
+            if (HttpContext.Features.Get<IHttpSendFileFeature>() != null)
```

Comment:
This is an interesting one. It means middleware that may provide send file as a shim won't get responses cached.